### PR TITLE
Adding support for aws proxy for deployments not having direct intern…

### PR DESCRIPTION
**Pull request by ajaijoy-bmc**
…et access
1. Updated clouddriver.yml with comments for adding proxy params including windows ntlm support.
2. Updated amazon credentials classes for the same.
3. Updated junits and ran junits.

### CODE REVIEW

**Comment by ajaijoy** [comment]
@ajordens  Please let me know your thoughts.

**Comment by cfieber** [comment]
Feels to me like the proxy configuration shouldn't be attached to the `NetflixAmazonCredentials` object

If you are managing many accounts, the proxy configuration for all of them should be the same (its about outbound connectivity from the host not the specific account being managed)

I would update the `AmazonClientProvider.Builder` to allow passing proxy configuration, update `AmazonClientProvider` to take proxy configuration as a constructor arg, and update `com.netflix.spinnaker.clouddriver.aws.AwsConfiguration` to inject the proxy configuration when it registers the `AmazonClientProvider` bean

You could make `AmazonCredentials.AWSProxy` an `@ConfigurationProperties` bound bean

**Comment by ajordens** [comment]
Attaching it one level higher in the hierarchy (aws vs aws.accounts) is a good call and is probably a simpler implementation / configuration.

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonClientProvider.java [hunk 1]
```diff
@@ -274,6 +275,45 @@ public AmazonCloudWatch getCloudWatch(NetflixAmazonCredentials amazonCredentials
     }
   }
 
+  private boolean isNTLMProxy(NetflixAmazonCredentials amazonCredentials){
+
+    boolean isNTLMProxy = false;
+
+    AmazonCredentials.AWSProxy proxy = amazonCredentials.getAWSProxy();
+
+    if(proxy != null){
```
Comment (ajordens) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonClientProvider.java]: Minor code formatting issues, spaces between `)` and `{`.

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonClientProvider.java [hunk 2]
```diff
@@ -274,6 +275,45 @@ public AmazonCloudWatch getCloudWatch(NetflixAmazonCredentials amazonCredentials
...
+        try{
+          Integer.parseInt(proxy.getProxyPort());
+        }catch(NumberFormatException nfe){
+           isProxy = false;
+        }
+
+
+    }
+
+    return isProxy;
```
Comment (ajordens) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonClientProvider.java]: `return proxy != null && proxy.isValid()'

Might be simpler if these methods were actually on `AWSProxy`.

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonClientProvider.java [hunk 3]
```diff
@@ -274,6 +275,45 @@ public AmazonCloudWatch getCloudWatch(NetflixAmazonCredentials amazonCredentials
...
+    AmazonCredentials.AWSProxy proxy = amazonCredentials.getAWSProxy();
+
+    if(proxy != null){
+
+      if(proxy.getProxyHost() != null && proxy.getProxyPort() != null && proxy.getProxyDomain() != null && proxy.getProxyWorkstation() != null )
+        isNTLMProxy = true;
+
+    }
+
+    return isNTLMProxy;
```
Comment (ajordens) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonClientProvider.java]: similar to below, a `proxy.isNTLM()` seems reasonable.

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonClientProvider.java [hunk 4]
```diff
@@ -281,6 +321,38 @@ public AmazonCloudWatch getCloudWatch(NetflixAmazonCredentials amazonCredentials
     ClientConfiguration clientConfiguration = new ClientConfiguration();
     clientConfiguration.setRetryPolicy(retryPolicy);
 
+    if(isProxyConfigMode(amazonCredentials)){
```
Comment (ajordens) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonClientProvider.java]:
```
if (proxy != null) {
   proxy.apply(clientConfiguration);
}
```

Or something along those lines might be more easily tested.

**Comment by ajordens** [comment]
Thanks for digging into this @ajaijoy-bmc!

**Comment by ajaijoy** [comment]
Thanks will do the changes.

**Comment by ajaijoy-bmc** [comment]
@cfieber  when i add @ConfigurationProperties to AmazonCredentials$AWSProxy without @Bean it says no bean found, @Bean doesnt compile. Any ideas?

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonClientProvider.java (hunk 1 above)
Comment (ajaijoy) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonClientProvider.java]: sure will fix.

File: clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonClientProvider.java (hunk 2 above)
Comment (ajaijoy) [inline clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/security/AmazonClientProvider.java]: agree, will move them to AWSProxy

**Comment by ajaijoy-bmc** [comment]
Closing this pull request and creating new pull request since i changed the design in the second attempt.

**Pull request closed without merging**